### PR TITLE
Fetch handles and external ids

### PR DESCRIPTION
**Pull request by jebbinSCL**

### CODE REVIEW

File: Imap/Program.fs [hunk 1]
```diff
@@ -1,18 +1,25 @@
...
 let main _ = 
-
+    setupLogging()
     let intervalSeconds = 60
     let interval = intervalSeconds * 1000
     let client = ZohoClient.getConnectedClient()
     let inbox = ZohoClient.openInbox client
 
+    let creds = Credentials.credentials
+
```
Comment (pallenSL) [inline Imap/Program.fs]: Is this used?

Comment (jebbinSCL) [inline Imap/Program.fs]: Not yet. taking it out for now.

File: Imap/App.config [hunk 2]
```diff
@@ -7,5 +7,11 @@
     <!--If you haven't created a Credentials config, go to Credentials.fs for details. EmailAccount and Password must be set-->
     <add key="EmailAccount" value=""/>
     <add key="Password" value =""/>
+
+    <add key="CommsUrl" value="http://localhost:8080"/>
+    <add key="SessionsUrl" value="Http://localhost:9000"/>
```
Comment (csscottlogic) [inline Imap/App.config]: Http?

File: Imap/CommsClient.fs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+open System
+open System.Net
+open System.Net.Http
+open System.Configuration
+open Newtonsoft.Json
+
+let url = 
+    let configUrl = ConfigurationManager.AppSettings.Item("CommsUrl")
+    if String.IsNullOrWhiteSpace configUrl then
+        failwith "Url for comms service is missing from configuration. Add url to procede."
```
Comment (csscottlogic) [inline Imap/CommsClient.fs]: proceed

File: Imap/HandlesClient.fs [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+open System.Net
+open System.Net.Http
+open System.Configuration
+open Newtonsoft.Json
+open Entities
+
+let url = 
+    let configUrl = ConfigurationManager.AppSettings.Item("SessionsUrl")
+    if String.IsNullOrWhiteSpace configUrl then
+        failwith "Url for comms service is missing from configuration. Add url to procede."
```
Comment (csscottlogic) [inline Imap/HandlesClient.fs]: proceed

File: Imap/ZohoClient.fs [hunk 5]
```diff
@@ -22,4 +27,10 @@ let getAllMessages (folder : IMailFolder) =
     folder.Search(SearchQuery.All)
     |> Seq.map (fun i-> folder.GetMessage(i))
 
+let getNewMessages (folder : IMailFolder) (knownIds : string[]) =
+    let summaries = folder.Fetch(0,-1,MessageSummaryItems.Envelope |||  MessageSummaryItems.UniqueId)
```
Comment (csscottlogic) [inline Imap/ZohoClient.fs]: Spacing? I think we agreed to have spaces after commas. Also, I think there are two after |||.

File: Imap/Program.fs [hunk 6]
```diff
@@ -1,18 +1,23 @@
...
     let client = ZohoClient.getConnectedClient()
     let inbox = ZohoClient.openInbox client
 
+    let handles = HandlesClient.getAllHandles()
+
     while true do 
         printfn "Starting Polling at %A" DateTime.UtcNow
-        ZohoClient.getAllMessages inbox
-        |> Seq.map Mapper.mimeMessageToEntity
+        let knownExternalIds = [||]
```
Comment (csscottlogic) [inline Imap/Program.fs]: We don't need this at the moment?

**Comment by csscottlogic** [comment]
CommsClient and HandlesClient have a lot of commonality?

File: Imap/Credentials.fs [hunk 7]
```diff
@@ -25,3 +25,5 @@ let getCredentials() =
     if String.IsNullOrWhiteSpace creds.Email  || String.IsNullOrWhiteSpace creds.Password then
```
Comment (csscottlogic) [inline Imap/Credentials.fs]: Spacing. Are there two spaces in front of ||?

File: Imap/ZohoClient.fs [hunk 8]
```diff
@@ -22,4 +27,10 @@ let getAllMessages (folder : IMailFolder) =
...
+    summaries
+    |> Seq.filter(fun x -> not <| Array.contains x.Envelope.MessageId knownIds)
```
Comment (csscottlogic) [inline Imap/ZohoClient.fs]: Why implement this at the moment?

File: Imap/Program.fs (hunk 6 above)
Comment (jebbinSCL) [inline Imap/Program.fs]: It'll actually be passed in in the next pull request, but at the moment it is just used for the test within getNewMessages.

File: Imap/Mapper.fs [hunk 9]
```diff
@@ -1,14 +1,36 @@
...
+
+let convertToISO8601 (datetime : DateTimeOffset) =
+    datetime.ToString("yyyy-MM-ddTHH\:mm\:ss\Z")
+
+let getProfileId (handles : HandleEntity []) email =
+    handles 
+    |> Seq.choose(fun x -> if x.Identifier = email then Some x.ProfileId else None)
+    |> Seq.tryHead
+
+let tryMapToEntity (handles : HandleEntity [])  (message : MimeMessage ) =
```
Comment (csscottlogic) [inline Imap/Mapper.fs]: We have HandleEntity and CorrespondenceItemEntity. ```Mapper.tryMapToEntity``` doesn't really tell me what's going on?

File: Imap/Mapper.fs [hunk 10]
```diff
@@ -1,14 +1,36 @@
...
+    |> Seq.choose(fun x -> if x.Identifier = email then Some x.ProfileId else None)
+    |> Seq.tryHead
+
+let tryMapToEntity (handles : HandleEntity [])  (message : MimeMessage ) =  
+    let senderHandle = message.From |> firstEmail
+    let receiverHandle = message.To |> firstEmail
+
+    match getProfileId handles senderHandle with
+    | None -> None
+    | Some senderId->
```
Comment (csscottlogic) [inline Imap/Mapper.fs]: Space before ->

File: Imap/ZohoClient.fs (hunk 8 above)
Comment (jebbinSCL) [inline Imap/ZohoClient.fs]: I'll take it out until the next pull request

File: Imap/CommsClient.fs [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+    else 
+        configUrl
+
+let getKnownExternalIds email = 
+    use client = new HttpClient()
+    let idsUrl = sprintf "%s/ExternalIds/?handle=%s" url email
```
Comment (pallenSL) [inline Imap/CommsClient.fs]: "ExternalIds/?handle".  I cannot trigger it to test but "/?" looks wrong.

File: Imap/InMemoryRepository.fs [hunk 12]
```diff
@@ -1,13 +1,14 @@
...
+let addIfNew (message : CorrespondenceItemEntity) =
     if store.ContainsKey(message.Id) then
         printfn "Found message with id %s. Not adding to store" message.Id 
         message.Id
     else 
         printfn "Storing message with id %s" message.Id
         store.Add(message.Id, message)
-        message.Id
\ No newline at end of file
+        message.Id
```
Comment (csscottlogic) [inline Imap/InMemoryRepository.fs]: Both branches of this if return ```message.Id```, so we can de-indent it by one level.

**Comment by jebbinSCL** [comment]
I'll cleanup the common code between Comms client and handles client in the next pull request.

**Comment by pallenSL** [comment]
Ok.
:ship:

**Pull request merged**